genimtools/src/igd/create.rs

```diff
@@ -0,0 +1,372 @@
+        .expect("File list path is required");
+
+    //Initialize IGD into Memory
+    let mut igd = igd_t::new();
+
+    igd.gType = 1;
+    igd.nbp = 16384; // from og code tile_size = 16384;
+    igd.nctg = 0;
+    igd.mctg = 32;
+    igd.total=0;
+
+    //Check that file path exists and get number of files
+    let mut all_bed_files: Vec<PathBuf>  = Vec::new();
+    //let mut all_bed_buffers = Vec::new();
+
+    let mut ix = 0;
+    let (mut start, mut end) = (0,0);
+
+    ///--------------------
+    /// Check each file and only keep the validated BED files
+    ///
+    /// -------------------
+
+    for entry in fs::read_dir(filelist).unwrap() {
+
+        // For now only take .bed files
+        if let Some(extension) = entry.as_ref().unwrap().path().extension() {
+
+            if extension != consts::FILE_EXTENSION.trim_start_matches('.') {
+                continue;
+            }
+        }
+
+        let entry = entry.unwrap();
+        let file_type = entry.file_type().unwrap();
+
+        if file_type.is_file() {
+
+            // open bed file
+            // TODO original code uses gzopen (I assume for .gz files?)
+            let file = File::open(entry.path()).unwrap();
+
+            let mut reader = BufReader::new(file);
+
+            /// Read the very first line and see if it meets our criteria
+            /// MUST USE by_ref() otherwise borrow checker won't let code compile
+            /// ALSO bec careful to call by_ref() BEFORE .lines()
+            ///
+            let first_line = reader.by_ref().lines().next().unwrap().expect("expect");
+            let mut lines = reader.lines();
+
+            // TODO Better name for og function?
+            // TODO parse_bed -> parse_bed_file_line
+            let ctg = parse_bed(&first_line, &mut start, &mut end);
+            // if it parses, add it to collected lines, increment ix
+            match ctg{
+
+                Some(ctg) =>{
+                    //all_bed_files.push(entry.path());
+                    //all_bed_files.push(line);
+                    //all_bed_buffers.push(lines);
+                    all_bed_files.push(entry.path());
+                    ix +=1;
+                } ,
+                None => continue,
+            }
+
+        }
+    }
+
+    //println!("ALL PARSED Lines from BED FILES:\n{:?}", all_bed_files);
+
+    let n_files = ix;//all_bed_files.len();
+
+    println!("Number of Bed Files found:\n{}", n_files);
+
+    //Check that there is more than 0 files?
+
+    //Prep memory allocation in a Rust-like manner
+    // TODO original code checks that the bed file can be parsed BEFORE memory allocation
+    // TODO but then re-parses the bed file again later.
+    // TODO use something like avg.shrink_to_fit(); after we've collected all the files?
+    // og C code:
+    //    int32_t *nr = calloc(n_files, sizeof(int32_t));
+    //     double *avg = calloc(n_files, sizeof(double));
+    let mut avg: Vec<i32> = Vec::with_capacity(n_files);
```

Comment:
We actually discussed this (partially) yesterday! I ended up having to use the `vec!` macro in some instances because `Vec::with_capacity` was not creating vectors of desired _lengths_. It was simply allocating memory for how large the vector is _expected_ to be come at run time:

> The capacity of a vector is the amount of space allocated for any future elements that will be added onto the vector. This is not to be confused with the length of a vector, which specifies the number of actual elements within the vector. If a vector's length exceeds its capacity, its capacity will automatically be increased, but its elements will have to be reallocated.
> 
> For example, a vector with capacity 10 and length 0 would be an empty vector with space for 10 more elements. Pushing 10 or fewer elements onto the vector will not change its capacity or cause reallocation to occur. However, if the vector's length is increased to 11, it will have to reallocate, which can be slow. For this reason, it is recommended to use Vec::with_capacity whenever possible to specify how big the vector is expected to get.

[link to info](https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/std/vec/struct.Vec.html#method.with_capacity)